framework/include/base/Assembly.h

```diff
@@ -713,57 +779,168 @@ class Assembly
    *
    * Note that this will also clear the cache.
    */
-  void addCachedJacobian(SparseMatrix<Number> & jacobian);
-
   void addCachedJacobian();
 
+  /**
+   * Same as addCachedJacobian but deprecated.
+   */
+  void addCachedJacobian(SparseMatrix<Number> & jacobian);
+
+  /**
+   * Get local residual block for a variable and a tag.
+   */
   DenseVector<Number> & residualBlock(unsigned int var_num, TagID tag_id = 0)
   {
     return _sub_Re[tag_id][var_num];
   }
 
+  /**
+   * Get local neighbor residual block for a variable and a tag.
+   */
   DenseVector<Number> & residualBlockNeighbor(unsigned int var_num, TagID tag_id = 0)
   {
     return _sub_Rn[tag_id][var_num];
   }
 
+  /**
+   * Get residual block for lower.
+   */
   DenseVector<Number> & residualBlockLower(unsigned int var_num, TagID tag_id = 0)
   {
     return _sub_Rl[tag_id][var_num];
   }
 
-  DenseMatrix<Number> & jacobianBlock(unsigned int ivar, unsigned int jvar, TagID tag = 0);
+  /**
+   * Get local Jacobian block for a pair of variables and a tag.
+   */
+  DenseMatrix<Number> & jacobianBlock(unsigned int ivar, unsigned int jvar, TagID tag = 0)
+  {
+    _jacobian_block_used[tag][ivar][jvar] = 1;
+    return _sub_Kee[tag][ivar][_block_diagonal_matrix ? 0 : jvar];
+  }
 
-  DenseMatrix<Number> & jacobianBlockNonlocal(unsigned int ivar, unsigned int jvar, TagID tag = 0);
+  /**
+   * Get local Jacobian block from non-local contribution for a pair of variables and a tag.
+   */
+  DenseMatrix<Number> & jacobianBlockNonlocal(unsigned int ivar, unsigned int jvar, TagID tag = 0)
+  {
+    _jacobian_block_nonlocal_used[tag][ivar][jvar] = 1;
+    return _sub_Keg[tag][ivar][_block_diagonal_matrix ? 0 : jvar];
+  }
+
+  /**
+   * Get local Jacobian block of a DG Jacobian type for a pair of variables and a tag.
+   */
   DenseMatrix<Number> & jacobianBlockNeighbor(Moose::DGJacobianType type,
                                               unsigned int ivar,
                                               unsigned int jvar,
-                                              TagID tag = 0);
+                                              TagID tag = 0)
+  {
```

Comment:
A small comment. I think compiler will inline all functions implemented in header file. If we move it to source, we might lose this. I guess I can add inline keyword for this function. Since this function is called only few places in the source file, I'd like it to be inlined.

Reply:
It will inline them if it thinks its best to inline them yes; it's more of a suggestion.

So since this is a public function that is used outside of the class and you want it inline, then take this out of the class definition, but keep it in the header file with the `inline` keyword.